src/log.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <iostream>
+#include <sstream>
+
+#include "bpftrace.h"
+
+namespace bpftrace {
+
+// clang-format off
+enum class LogType
+{
+  debug,
+  info,
+  warning,
+  error
+};
+// clang-format on
+
+class Log
+{
+public:
+  static Log& get();
+  void take_input(LogType type,
+                  const location& loc,
+                  bool with_loc,
+                  std::ostream& out,
+                  std::string&& input);
+  inline void set_source(std::string filename, std::string source)
+  {
+    src_ = source;
+    filename_ = filename;
+  }
+  inline const std::string& get_source()
+  {
+    return src_;
+  }
+  const std::string get_source_line(unsigned int n);
+
+  // Can only construct with get()
+  Log(const Log& other) = delete;
+  Log& operator=(const Log& other) = delete;
+
+private:
+  Log() = default;
+  ~Log() = default;
+  std::string src_;
+  std::string filename_;
+  void log_with_location(LogType,
+                         const location&,
+                         std::ostream&,
+                         const std::string&);
+};
+
+class LogStream
+{
+public:
+  LogStream(const std::string file,
```

Comment:
Nit: doesn't matter in this case (compiler will probably RVO), but IIUC const-ref is still superior in case the api gets used differently later.
```suggestion
  LogStream(const std::string file,
```